<a href="https://colab.research.google.com/github/weedge/doraemon-nb/blob/main/redis_vss_colab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import requests
import json

url = 'https://raw.githubusercontent.com/bsbodden/redis_vss_getting_started/main/data/bikes.json'
response = requests.get(url)
bikes = json.loads(response.text)

In [ ]:
%pip -q install pandas
import pandas as pd

pd.DataFrame(bikes)


model       brand  price                 type  \
0                 Jigger     Velorim    270           Kids bikes   
1              Hillcraft       Bicyk   1200  Kids Mountain Bikes   
2            Chook air 5        Nord    815  Kids Mountain Bikes   
3                Eva 291         Eva   3400       Mountain Bikes   
4                 Kahuna  Noka Bikes   3200       Mountain Bikes   
5          XBN 2.1 Alloy    Breakout    810           Road Bikes   
6               WattBike  ScramBikes   2300               eBikes   
7   Soothe Electric bike   Peaknetic   1950               eBikes   
8                  Secto   Peaknetic    430       Commuter bikes   
9                 Summit       nHill   1200        Mountain Bike   
10           ThrillCycle   BikeShind    815       Commuter Bikes   

                                          specs  \
0     {'material': 'aluminium', 'weight': '10'}   
1        {'material': 'carbon', 'weight': '11'}   
2        {'material': 'alloy', 'weight': '9.1'}   
3       {'material': 'carbon', 'weight': '9.1'}   
4        {'material': 'alloy', 'weight': '9.8'}   
5        {'material': 'alloy', 'weight': '7.2'}   
6         {'material': 'alloy', 'weight': '15'}   
7       {'material': 'alloy', 'weight': '14.7'}   
8   {'material': 'aluminium', 'weight': '10.0'}   
9       {'material': 'alloy', 'weight': '11.3'}   
10      {'material': 'alloy', 'weight': '12.7'}   

                                          description  
0   Small and powerful, the Jigger is the best rid...  
1   Kids want to ride with as little weight as pos...  
2   The Chook Air 5  gives kids aged six years and...  
3   The sister company to Nord, Eva launched in 20...  
4   Whether you want to try your hand at XC racing...  
5   The XBN 2.1 Alloy is our entry-level road bike...  
6   The WattBike is the best e-bike for people who...  
7   The Soothe is an everyday electric bike, from ...  
8   If you struggle with stiff fingers or a kinked...  
9   This budget mountain bike from nHill performs ...  
10  An artsy,  retro-inspired bicycle that’s as fu...

In [6]:
print(json.dumps(bikes[0], indent=2))


{
  "model": "Jigger",
  "brand": "Velorim",
  "price": 270,
  "type": "Kids bikes",
  "specs": {
    "material": "aluminium",
    "weight": "10"
  },
  "description": "Small and powerful, the Jigger is the best ride for the smallest of tikes! This is the tiniest kids\u2019 pedal bike on the market available without a coaster brake, the Jigger is the vehicle of choice for the rare tenacious little rider raring to go. We say rare because this smokin\u2019 little bike is not ideal for a nervous first-time rider, but it\u2019s a true giddy up for a true speedster. The Jigger is a 12 inch lightweight kids bicycle and it will meet your little one\u2019s need for speed. It\u2019s a single speed bike that makes learning to pump pedals simple and intuitive. It even has  a handle in the bottom of the saddle so you can easily help your child during training!  The Jigger is among the most lightweight children\u2019s bikes on the planet. It is designed so that 2-3 year-olds fit comfortably in a mo

In [7]:
%pip install -U -q sentence-transformers


In [8]:
from textwrap import TextWrapper

sample_description = bikes[0]['description']
TextWrapper(width=120).wrap(sample_description)


['Small and powerful, the Jigger is the best ride for the smallest of tikes! This is the tiniest kids’ pedal bike on the',
 'market available without a coaster brake, the Jigger is the vehicle of choice for the rare tenacious little rider raring',
 'to go. We say rare because this smokin’ little bike is not ideal for a nervous first-time rider, but it’s a true giddy',
 'up for a true speedster. The Jigger is a 12 inch lightweight kids bicycle and it will meet your little one’s need for',
 'speed. It’s a single speed bike that makes learning to pump pedals simple and intuitive. It even has  a handle in the',
 'bottom of the saddle so you can easily help your child during training!  The Jigger is among the most lightweight',
 'children’s bikes on the planet. It is designed so that 2-3 year-olds fit comfortably in a molded ride position that',
 'allows for efficient riding, balanced handling and agility. The Jigger’s frame design and gears work together so your',
 'buddingbiker can stand 

In [10]:
from sentence_transformers import SentenceTransformer

embedder = SentenceTransformer('msmarco-distilbert-base-v4')

In [11]:
embedding = embedder.encode(sample_description)

In [13]:
print(embedding.tolist()[:5])

[0.20076490938663483, -0.13000741600990295, 0.3081614077091217, 0.20627954602241516, -0.36923569440841675]


In [14]:
VECTOR_DIMENSION = len(embedding)
VECTOR_DIMENSION

768

In [ ]:
%apt-get install libssl1.1

In [21]:
in_colab = 'google.colab' in str(get_ipython())
if in_colab:
  !echo "deb http://security.ubuntu.com/ubuntu focal-security main" | sudo tee /etc/apt/sources.list.d/focal-security.list
  !apt-get update && \
    apt-get install libssl1.1
  !curl -fsSL https://packages.redis.io/redis-stack/redis-stack-server-6.2.6-v7.focal.x86_64.tar.gz -o redis-stack-server.tar.gz
  !tar -xvf redis-stack-server.tar.gz
  !./redis-stack-server-6.2.6-v7/bin/redis-stack-server --daemonize yes


deb http://security.ubuntu.com/ubuntu focal-security main
Hit:1 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:2 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [119 kB]
Get:3 http://security.ubuntu.com/ubuntu jammy-security InRelease [110 kB]
Get:4 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,626 B]
Get:5 http://security.ubuntu.com/ubuntu focal-security InRelease [114 kB]
Get:6 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [109 kB]
Hit:7 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease
Get:8 http://security.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [1,059 kB]
Hit:9 https://ppa.launchpadcontent.net/c2d4u.team/c2d4u4.0+/ubuntu jammy InRelease
Get:10 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [962 kB]
Get:11 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [1,230 kB]
Get:12 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Pa

In [18]:
%pip install -q redis

In [22]:
import redis
client = redis.Redis(host = 'localhost', port=6379, decode_responses=True)

In [23]:
client.ping()

True

In [24]:
pipeline = client.pipeline()

for i, bike in enumerate(bikes, start=1):
    redis_key = f'bikes:{i:03}'
    pipeline.json().set(redis_key, '$', bike)

pipeline.execute()

[True, True, True, True, True, True, True, True, True, True, True]

In [25]:
client.json().get('bikes:010', '$.model')

['Summit']

In [26]:
keys = sorted(client.keys('bikes:*'))
print(keys)

['bikes:001', 'bikes:002', 'bikes:003', 'bikes:004', 'bikes:005', 'bikes:006', 'bikes:007', 'bikes:008', 'bikes:009', 'bikes:010', 'bikes:011']


In [27]:
import numpy as np

descriptions = client.json().mget(keys, '$.description')
descriptions = [item for sublist in descriptions for item in sublist]
embeddings = embedder.encode(descriptions).astype(np.float32).tolist()


In [28]:
pipeline = client.pipeline()

for key, embedding in zip(keys, embeddings):
    pipeline.json().set(key, '$.description_embeddings', embedding)

pipeline.execute()

[True, True, True, True, True, True, True, True, True, True, True]

In [29]:
print(json.dumps(client.json().get('bikes:010'), indent=2))

{
  "model": "Summit",
  "brand": "nHill",
  "price": 1200,
  "type": "Mountain Bike",
  "specs": {
    "material": "alloy",
    "weight": "11.3"
  },
  "description": "This budget mountain bike from nHill performs well both on bike paths and on the trail. The fork with 100mm of travel absorbs rough terrain. Fat Kenda Booster tires give you grip in corners and on wet trails. The Shimano Tourney drivetrain offered enough gears for finding a comfortable pace to ride uphill, and the Tektro hydraulic disc brakes break smoothly. Whether you want an affordable bike that you can take to work, but also take trail riding on the weekends or you\u2019re just after a stable, comfortable ride for the bike path, the Summit gives a good value for money.",
  "description_embeddings": [
    -0.5381144285202026,
    -0.49465906620025635,
    -0.02517692558467388,
    0.6540352702140808,
    -0.06241397559642792,
    -0.6898808479309082,
    -0.5430218577384949,
    -0.5903494954109192,
    0.50613242387

When storing a vector embedding as part of a JSON datatype, the embedding is stored as a JSON array, in our case, under the field description_embeddings as shown.

Making the bikes collection searchable
Redis Stack provides a powerful search engine (RediSearch) that introduces commands to create and maintain search indices for both collections of HASHES and JSON documents.

To create a search index for the bikes collection, we'll use the FT.CREATE command:

```
1️⃣ FT.CREATE idx:bikes_vss ON JSON
2️⃣  PREFIX 1 bikes: SCORE 1.0
3️⃣  SCHEMA
4️⃣    $.model TEXT WEIGHT 1.0 NOSTEM
5️⃣    $.brand TEXT WEIGHT 1.0 NOSTEM
6️⃣    $.price NUMERIC
7️⃣    $.type TAG SEPARATOR ","
8️⃣    $.description AS description TEXT WEIGHT 1.0
9️⃣    $.description_embeddings AS vector VECTOR FLAT 6 TYPE FLOAT32 DIM 768 DISTANCE_METRIC COSINE
```
There is a lot to unpack here; let's take it from the top:

1️⃣ We start by specifying the name of the index; idx:bikes indexing keys of type JSON.

2️⃣ The keys being indexed are found using the bikes: key prefix.

3️⃣ The SCHEMA keyword marks the beginning of the schema field definitions.

4️⃣ Declares that field in the JSON document at the JSONPath $.model will be indexed as a TEXT field, allowing full-text search queries (disabling stemming).

5️⃣ The $.brand field will also be treated as a TEXT schema field.

6️⃣ The $.price field will be indexed as a NUMERIC allowing numeric range queries.

7️⃣ The $.type field will be indexed as a TAG field. Tag fields allow exact-match queries, and are suitable for categorical values.

8️⃣ The $.description field will also be indexed as a TEXT field.

9️⃣ Finally, the vector embeddings in $.description_embeddings are indexed as a VECTOR field and assigned to the alias vector.

Let's break down the VECTOR schema field definition to better understand the inner workings of Vector Similarity in Redis:

FLAT: Specifies the indexing method, which can be FLAT or HNSW. FLAT (brute-force indexing) provides exact results but at a higher computational cost, while HNSW (Hierarchical Navigable Small World) is a more efficient method that provides approximate results with lower computational overhead.
TYPE: Set to FLOAT32. Current supported types are FLOAT32 and FLOAT64.
DIM: The length or dimension of our embeddings, which we determined previosly to be 768.

DISTANCE_METRIC: One of L2, IP, COSINE.

L2 stands for "Euclidean distance"; a straight-line distance between the vectors. Preferred when the absolute differences, including magnitude, matter most.

IP stands for "Inner Product"; IP measures the projection of one vector onto another. It emphasizes the angle between vectors rather than their absolute positions in the vector space.

COSINE stands for "Cosine Similarity"; a normalized form of inner product. This metric measures only the angle between two vectors, making it magnitude-independent.

For our querying purposes, the direction of the vectors carry more meaning (indicating semantic similarity), and the magnitude is largely influenced by the length of the documents, therefore COSINE similarity is chosen. Also, our chosen embedding model is fine-tuned for Cosine Similarity.

The Python code below creates the Redis Search Index for the bikes collection equivalent to the previous raw FT.CREATE command:

In [30]:
from redis.commands.search.field import TagField, TextField, NumericField, VectorField
from redis.commands.search.indexDefinition import IndexDefinition, IndexType
from redis.commands.search.query import Query

INDEX_NAME = 'idx:bikes_vss'
DOC_PREFIX = 'bikes:'

try:
    # check to see if index exists
    client.ft(INDEX_NAME).info()
    print('Index already exists!')
except:
    # schema
    schema = (
        TextField('$.model', no_stem=True, as_name='model'),
        TextField('$.brand', no_stem=True, as_name='brand'),
        NumericField('$.price', as_name='price'),
        TagField('$.type', as_name='type'),
        TextField('$.description', as_name='description'),
        VectorField('$.description_embeddings',
            'FLAT', {
                'TYPE': 'FLOAT32',
                'DIM': VECTOR_DIMENSION,
                'DISTANCE_METRIC': 'COSINE',
            },  as_name='vector'
        ),
    )

    # index Definition
    definition = IndexDefinition(prefix=[DOC_PREFIX], index_type=IndexType.JSON)

    # create Index
    client.ft(INDEX_NAME).create_index(fields=schema, definition=definition)


In [31]:
info = client.ft(INDEX_NAME).info()

num_docs = info['num_docs']
indexing_failures = info['hash_indexing_failures']
total_indexing_time = info['total_indexing_time']
percent_indexed = int(info['percent_indexed']) * 100


print(f"{num_docs} documents ({percent_indexed} percent) indexed with {indexing_failures} failures in {float(total_indexing_time):.2f} milliseconds")


11 documents (100 percent) indexed with 0 failures in 4.78 milliseconds


Structured Data Searches with Redis

The index idx:bikes_vss indexes the structured fields of our JSON documents model, brand, price, and type. It also indexes the unstructured free-form text description and the generated embeddings in description_embeddings. Before we dive deeper into Vector Similarity Search, we need to understand the basics of querying a Redis index. The Redis command of interest is FT.SEARCH. Like a SQL select statement, an FT.SEARCH invocation can be as simple or as complex as needed.

Let's try simple queries that give enough context to complete our VSS examples. For example, to retrieve all bikes where the brand is Peaknetic, we can use the following command:

`FT.SEARCH idx:bikes_vss '@brand:Peaknetic'`

The command will return all matching documents. With the inclusion of the vector embeddings, that's a little too verbose. If we wanted to only return specific fields from our JSON documents, for example, the document id, the brand, model and price, we could use:

`FT.SEARCH idx:bikes_vss '@brand:Peaknetic' RETURN 4 id, brand, model, price`

In the query, we are searching against a schema field of type TEXT. The equivalent Python code is:



In [32]:
query = (
    Query('@brand:Peaknetic @price:[0 1000]').return_fields('id', 'brand', 'model', 'price')
)
client.ft(INDEX_NAME).search(query).docs


[Document {'id': 'bikes:009', 'payload': None, 'brand': 'Peaknetic', 'model': 'Secto', 'price': '430'}]

Semantic Searching with Vector Similarity Search

Now that the bikes collection is stored and properly indexed in Redis, we want to query them using short query prompts. Let's put our queries in a list so we can execute them in bulk:



In [33]:
queries = [
    'Bike for small kids',
    'Best Mountain bikes for kids',
    'Cheap Mountain bike for kids',
    'Female specific mountain bike',
    'Road bike for beginners',
    'Commuter bike for people over 60',
    'Comfortable commuter bike',
    'Good bike for college students',
    'Mountain bike for beginners',
    'Vintage bike',
    'Comfortable city bike'
]


In [34]:
#We need to encode the query prompts to query the database using VSS. Just like we did with the descriptions of the bikes,
#we'll use the SentenceTransformers model to encode the queries:
encoded_queries = embedder.encode(queries)
len(encoded_queries)


11

## Constructing a "Pure KNN" VSS Query
We'll start with a K-nearest neighbors (KNN) query. KNN is a foundational algorithm used in vector similarity search, where the goal is to find the most similar items to a given query item. Using the chosen distance metric, the KNN algorithm calculates the distance between the query vector and each vector in the database. It then returns the 'K' items with the smallest distances to the query vector. These are considered to be the most similar items.

The syntax for vector similarity KNN queries is (\*)=>[vector_similarity_query>] where the (\*) (the \* meaning all) is the filter query for the search engine. That way, one can reduce the search space by filtering the collection on which the KNN algorithm operates.

The \$query_vector represents the query parameter we'll use to pass the vectorized query prompt.

The results will be filtered by vector_score, which is a field derived from the name of the field indexed as a Vector by appending _score to it, in our case, vector (the alias for \$.description_embeddings).

Our query will return the vector_score, the id of the match documents, and the \$.brand, \$.model, and \$.description.

Finally, to utilize a vector similarity query with the FT.SEARCH command, we must specify DIALECT 2 or greater.


In [35]:
query = (
    Query('(*)=>[KNN 3 @vector $query_vector AS vector_score]')
     .sort_by('vector_score')
     .return_fields('vector_score', 'id', 'brand', 'model', 'description')
     .dialect(2)
)


We pass the vectorized query as \$query_vector to the search function to execute the query. The following code shows an example of creating a NumPy array from a vectorized query prompt (encoded_query) as a single precision float array and converting it into a compact, byte-level representation that we can pass as a Redis parameter:

```python
client.ft(INDEX_NAME).search(query, { 'query_vector': np.array(encoded_query, dtype=np.float32).tobytes() }).docs
```

With the template for the query in place, we can use a bit of Python code to execute all query prompts in a loop, passing the vectorized query prompts. Notice that for each result we calculate the vector_score as 1 - doc.

vector_score, since we use cosine "distance" as the metric, the items with the smallest "distance" are closer and therefore more similar to our query.

We will then loop over the matched documents and create a list of results that will help up make a lovely Pandas table to visualize the results:



In [37]:
from IPython.display import display, HTML

def create_query_table(query, queries, encoded_queries, extra_params = {}):
    results_list = []
    for i, encoded_query in enumerate(encoded_queries):
        result_docs = client.ft(INDEX_NAME).search(query, { 'query_vector': np.array(encoded_query, dtype=np.float32).tobytes() } | extra_params).docs
        for doc in result_docs:
            vector_score = round(1 - float(doc.vector_score), 2)
            results_list.append({
                'query': queries[i],
                'score': vector_score,
                'id': doc.id,
                'brand': doc.brand,
                'model': doc.model,
                'description': doc.description
            })

    # Pretty-print the table
    queries_table = pd.DataFrame(results_list)
    queries_table.sort_values(by=['query', 'score'], ascending=[True, False], inplace=True)
    queries_table['query'] = queries_table.groupby('query')['query'].transform(lambda x: [x.iloc[0]] + ['']*(len(x)-1))
    queries_table['description'] = queries_table['description'].apply(lambda x: (x[:497] + '...') if len(x) > 500 else x)
    html = queries_table.to_html(index=False)
    display(HTML(html))

The query results show the individual queries' top 3 matches (our K parameter) along with the bike's id, brand, and model for each query. For example, for the query "Best Mountain bikes for kids", the highest similarity score (0.54) and therefore the closest match was the 'Nord' brand 'Chook air 5' bike model, described as:

> "The Chook Air 5 gives kids aged six years and older a durable and uberlight mountain bike for their first experience on tracks and easy cruising through forests and fields. The lower top tube makes it easy to mount and dismount in any situation, giving your kids greater safety on the trails. The Chook Air 5 is the perfect intro to mountain biking."

From the description, we gather that this bike is an excellent match for younger children, and the MS MARCO model-generated embeddings seem to have captured the semantics of the description accurately.



In [38]:
create_query_table(query, queries, encoded_queries)


In [39]:
hybrid_query = (
    Query('(@brand:Peaknetic)=>[KNN 3 @vector $query_vector AS vector_score]')
     .sort_by('vector_score')
     .return_fields('vector_score', 'id', 'brand', 'model', 'description')
     .dialect(2)
)


In [40]:
create_query_table(hybrid_query, queries, encoded_queries)


## Creating a VSS Range Query
Range queries in Vector Similarity Search (VSS) involve retrieving items within a specific distance from a query vector. In this case, we consider "distance" to be the measure of similarity we've used to build our search indices; the smaller the distance, the more similar the items.

Let's say you want to find the bikes whose descriptions are within a certain distance from a query vector. We can use a range query to achieve this. For example, the query command to return the top 4 documents within a 0.55 radius of a vectorized query would be as follows:
```shell
1️⃣ FT.SEARCH idx:bikes_vss
2️⃣   @vector:[VECTOR_RANGE $range $query_vector]=>{$YIELD_DISTANCE_AS: vector_score}
3️⃣   SORTBY vector_score ASC
4️⃣   LIMIT 0 4
5️⃣   DIALECT 2
6️⃣   PARAMS 4 range 0.55 query_vector "\x9d|\x99>bV#\xbfm\x86\x8a\xbd\xa7~$?*...."
```

Where:

1️⃣ We use the FT.SEARCH command with our idx:bikes_vss.

2️⃣ and filter by the vector using the VECTOR_RANGE operator pasing the $range parameter, yield the vector distance between the vector field and the query result in a field named vector_score.

3️⃣ We sort the results by the yielded vector_score.

4️⃣ Limit the results to at most 4.

5️⃣ Once again, we set the RediSearch dialect to 2 to enable VSS functionality.

6️⃣ Finally we set the parameter values, range ($range) to 0.55 and the query_vector ($query_vector) to the encoded vectorized query.

The equivalent Python query definition is shown below:



In [41]:
range_query = (
    Query('@vector:[VECTOR_RANGE $range $query_vector]=>{$YIELD_DISTANCE_AS: vector_score}')
    .sort_by('vector_score')
    .return_fields('vector_score', 'id', 'brand', 'model', 'description')
    .paging(0, 4)
    .dialect(2)
)


In [42]:
create_query_table(range_query, queries[:1], encoded_queries[:1], {'range': 0.55})


## Wrapping Up
In this guide, we learned how Redis, using the Redis Stack distribution, provides powerful search capabilities over structured and unstructured data.

Redis support for vector data can enrich and enhance the user's search experience. Although we focused on generating embeddings for unstructured data, the vector similarity approach can equally be employed with structure data, as long as a suitable vector generation technique is used.

The references below can help you learn more about Redis search capabilities:
* https://redis.io/docs/stack/search/
* https://redis.io/docs/stack/search/indexing_json/